### PR TITLE
Stop at the first `NULL` argument when iterating `argv`

### LINKED ISSUE #105999: glibc `.init_array` usage for `std::env::args` is not safe
Currently in `library/std/src/sys/unix/args.rs`, Rust is hooking into the glibc `.init_array` extension for retrieving `argc`/`argv` even in case Rust does not manage `main()` and is e.g. just loaded into a process as a `cdylib`.

While this is great and convenient, it's unfortunately not implemented in a safe way. Various C commandline argument parsers are **modifying** `argv` while parsing, so the information that gets passed into `.init_array` might not be correct anymore.

Examples of such parsers are

  - glibc's own [`getopt`](https://www.gnu.org/software/libc/manual/html_node/Using-Getopt.html) parser, noting " The default is to permute the contents of argv while scanning it so that eventually all the non-options are at the end. This allows options to be given in any order, even with programs that were not written to expect this. "
  - GLib's [`GOptionContext`](https://docs.gtk.org/glib/struct.OptionContext.html) parser, which removes all already handled arguments from `argv` and only leaves others (removed ones are set to `NULL` and moved to the end).
  - Qt does the same thing in `QCoreApplication` in `QCoreApplicationPrivate::processCommandLineArguments()`. It shuffles around arguments in `argv` and removes (by setting to `nullptr`) arguments it handled already.

This means that at least `argc` can easily be too big, and Rust would read beyond the (new) end of `argv`. This caused crashes in [practice](https://gitlab.freedesktop.org/gstreamer/gst-plugins-rs/-/issues/280) because of calling `strlen()` on `NULL` when creating an `CStr` around such an "removed" argument.

Now the question is how this should be handled in Rust. I see three options here

  1. Remove the `.init_array` extension usage and handle glibc like all other libcs
  2. Copy all arguments in `.init_array`. This means everybody has to pay for that even if they don't use arguments, and this is theoretically still not safe if the shared library is loaded at the same time the `argv` array is modified and a partially written pointer value is read
  3. Handle `NULL` pointers in `argv` by skipping over them. This means we would lose the exact length information of the `args` iterator, and this is still theoretically not safe for the same reason as 2. See https://github.com/rust-lang/rust/pull/106001

This code was added in 2019 by https://github.com/rust-lang/rust/pull/66547 (CC @leo60228)

~~I'd be happy to provide a PR implementing either solution once there is some agreement how to move forward here.~~ I've created https://github.com/rust-lang/rust/pull/106001 as a proposed fix for this, which implements option 3.

### Meta

`rustc --version --verbose`:
```
rustc 1.66.0 (69f9c33d7 2022-12-12)
binary: rustc
commit-hash: 69f9c33d71c871fc16ac445211281c6e7a340943
commit-date: 2022-12-12
host: x86_64-unknown-linux-gnu
release: 1.66.0
LLVM version: 15.0.2
```

<details><summary>Backtrace</summary>
<p>

```
Program received signal SIGSEGV, Segmentation fault.
__strlen_evex () at ../sysdeps/x86_64/multiarch/strlen-evex.S:79
79		VPCMP	$0, (%rdi), %YMMZERO, %k0
Missing separate debuginfos, use: dnf debuginfo-install bzip2-libs-1.0.8-12.fc37.x86_64 elfutils-libelf-0.188-3.fc37.x86_64 elfutils-libs-0.188-3.fc37.x86_64 libblkid-2.38.1-1.fc37.x86_64 libmount-2.38.1-1.fc37.x86_64 libunwind-1.6.2-5.fc37.x86_64 libzstd-1.5.2-3.fc37.x86_64 pcre2-10.40-1.fc37.1.x86_64 xz-libs-5.2.5-10.fc37.x86_64 zlib-1.2.12-5.fc37.x86_64
(gdb) backtrace full
#0  __strlen_evex () at ../sysdeps/x86_64/multiarch/strlen-evex.S:79
#1  0x00007ffff7713281 in core::ffi::c_str::CStr::from_ptr () at library/core/src/ffi/c_str.rs:286
#2  std::sys::unix::args::imp::clone::{closure#0} () at library/std/src/sys/unix/args.rs:146
#3  core::iter::adapters::map::map_fold::{closure#0}<isize, std::ffi::os_str::OsString, (), std::sys::unix::args::imp::clone::{closure_env#0}, core::iter::traits::iterator::Iterator::for_each::call::{closure_env#0}<std::ffi::os_str::OsString, alloc::vec::spec_extend::{impl#1}::spec_extend::{closure_env#0}<std::ffi::os_str::OsString, core::iter::adapters::map::Map<core::ops::range::Range<isize>, std::sys::unix::args::imp::clone::{closure_env#0}>, alloc::alloc::Global>>> () at library/core/src/iter/adapters/map.rs:84
#4  core::iter::traits::iterator::Iterator::fold<core::ops::range::Range<isize>, (), core::iter::adapters::map::map_fold::{closure_env#0}<isize, std::ffi::os_str::OsString, (), std::sys::unix::args::imp::clone::{closure_env#0}, core::iter::traits::iterator::Iterator::for_each::call::{closure_env#0}<std::ffi::os_str::OsString, alloc::vec::spec_extend::{impl#1}::spec_extend::{closure_env#0}<std::ffi::os_str::OsString, core::iter::adapters::map::Map<core::ops::range::Range<isize>, std::sys::unix::args::imp::clone::{closure_env#0}>, alloc::alloc::Global>>>>
    () at library/core/src/iter/traits/iterator.rs:2414
#5  core::iter::adapters::map::{impl#2}::fold<std::ffi::os_str::OsString, core::ops::range::Range<isize>, std::sys::unix::args::imp::clone::{closure_env#0}, (), core::iter::traits::iterator::Iterator::for_each::call::{closure_env#0}<std::ffi::os_str::OsString, alloc::vec::spec_extend::{impl#1}::spec_extend::{closure_env#0}<std::ffi::os_str::OsString, core::iter::adapters::map::Map<core::ops::range::Range<isize>, std::sys::unix::args::imp::clone::{closure_env#0}>, alloc::alloc::Global>>> () at library/core/src/iter/adapters/map.rs:124
#6  core::iter::traits::iterator::Iterator::for_each<core::iter::adapters::map::Map<core::ops::range::Range<isize>, std::sys::unix::args::imp::clone::{closure_env#0}>, alloc::vec::spec_extend::{impl#1}::spec_extend::{closure_env#0}<std::ffi::os_str::OsString, core::iter::adapters::map::Map<core::ops::range::Range<isize>, std::sys::unix::args::imp::clone::{closure_env#0}>, alloc::alloc::Global>> () at library/core/src/iter/traits/iterator.rs:831
#7  alloc::vec::spec_extend::{impl#1}::spec_extend<std::ffi::os_str::OsString, core::iter::adapters::map::Map<core::ops::range::Range<isize>, std::sys::unix::args::imp::clone::{closure_env#0}>, alloc::alloc::Global> () at library/alloc/src/vec/spec_extend.rs:40
#8  alloc::vec::spec_from_iter_nested::{impl#1}::from_iter<std::ffi::os_str::OsString, core::iter::adapters::map::Map<core::ops::range::Range<isize>, std::sys::unix::args::imp::clone::{closure_env#0}>> () at library/alloc/src/vec/spec_from_iter_nested.rs:62
#9  alloc::vec::spec_from_iter::{impl#0}::from_iter<std::ffi::os_str::OsString, core::iter::adapters::map::Map<core::ops::range::Range<isize>, std::sys::unix::args::imp::clone::{closure_env#0}>> () at library/alloc/src/vec/spec_from_iter.rs:33
#10 alloc::vec::{impl#18}::from_iter<std::ffi::os_str::OsString, core::iter::adapters::map::Map<core::ops::range::Range<isize>, std::sys::unix::args::imp::clone::{closure_env#0}>> ()
    at library/alloc/src/vec/mod.rs:2757
#11 core::iter::traits::iterator::Iterator::collect<core::iter::adapters::map::Map<core::ops::range::Range<isize>, std::sys::unix::args::imp::clone::{closure_env#0}>, alloc::vec::Vec<std::ffi::os_str::OsString, alloc::alloc::Global>> () at library/core/src/iter/traits/iterator.rs:1836
#12 std::sys::unix::args::imp::clone () at library/std/src/sys/unix/args.rs:144
#13 std::sys::unix::args::imp::args () at library/std/src/sys/unix/args.rs:129
#14 std::sys::unix::args::args () at library/std/src/sys/unix/args.rs:19
#15 std::env::args_os () at library/std/src/env.rs:792
#16 0x00007ffff7713151 in std::env::args () at library/std/src/env.rs:757
[...]
```

Also

```
(gdb) info registers
rax            0x0                 0
rbx            0x8                 8
rcx            0x6b6e6973656b6166  7741240753840152934
rdx            0x8                 8
rsi            0x6b6e6973656b6166  7741240753840152934
rdi            0x0                 0
rbp            0x0                 0x0
rsp            0x7fffffffc398      0x7fffffffc398
r8             0x7ffff7cacce0      140737350651104
r9             0x40                64
r10            0x0                 0
r11            0x20                32
r12            0x55555556e890      93824992340112
r13            0x4                 4
r14            0x5                 5
r15            0x5555558748e0      93824995510496
rip            0x7ffff7c4093c      0x7ffff7c4093c <__strlen_evex+28>
```

</p>
</details>

### PR DESCRIPTION
**Pull request by sdroege**
Some C commandline parsers (e.g. GLib and Qt) are replacing already handled arguments in `argv` with `NULL` and move them to the end. That means that `argc` might be bigger than the actual number of non-`NULL` pointers in `argv` at this point.
    
To handle this we simply stop iterating at the first `NULL` argument.
    
`argv` is also guaranteed to be `NULL`-terminated so any non-`NULL` arguments after the first `NULL` can safely be ignored.

Fixes https://github.com/rust-lang/rust/issues/105999

### CODE REVIEW

File: library/std/src/sys/unix/args.rs [hunk 1]
```diff
@@ -142,9 +142,20 @@ mod imp {
...
-                    OsStringExt::from_vec(cstr.to_bytes().to_vec())
+                .filter_map(|i| {
+                    let ptr = *argv.offset(i) as *const libc::c_char;
+                    // Some C commandline parsers are replacing already handled arguments in `argv`
+                    // with `NULL` so it's necessary to skip over them here.
+                    //
+                    // We only worry about that on glibc as that is the only situation where Rust
+                    // is not in charge of `main()` but can still retrieve `argc`/`argv` via
+                    // `.init_array`.
+                    if cfg!(all(target_os = "linux", target_env = "gnu")) && ptr.is_null() {
```
Comment (bjorn3) [inline library/std/src/sys/unix/args.rs:153]: On other libc's someone may use a C commandline parser from their rust main function, right? In that case argv would still be modified after libstd stored argc/argv during it's pre-main initialization.

Comment (sdroege) [reply library/std/src/sys/unix/args.rs:153]: You would have to actually get access to the `argv` pointer (it's not passed into `main()`), and then pass it to a C function (via FFI, so `unsafe`).

I'm fine with doing the `NULL` check everywhere and always if we want to allow the above. That `NULL` check is not really expensive and commandline argument parsing is probably not a very performance critical aspect of any application where one additional `NULL` check per argument would hurt :)

Comment (bjorn3) [reply library/std/src/sys/unix/args.rs:153]: On macOS there is `_NSGetArgv()`. In fact libstd uses this function.

Comment (Nilstrieb) [reply library/std/src/sys/unix/args.rs:153]: We allocate each argument anyways, so this tiny check shouldn't matter. It's safer to just always do it.

Comment (sdroege) [reply library/std/src/sys/unix/args.rs]: Ok then :)

**Comment by sdroege** [comment]
Is there anything I can help to move this forward? It would be great if this wouldn't cause crashes anymore :sweat_smile:

**Comment by sdroege** [comment]
@joshtriplett As this basically implements what you suggested in the issues and this is assigned to you for review, is there anything I can help with to get this reviewed faster?

**Comment by RalfJung** [comment]
Might be worth re-rolling the reviewer dice.

r? libs

**Comment by ChrisDenton** [comment]
Code looks good to me. I happy to accept this given this is a potential safety issue (or at least this PR is better than crashing). And [josh's opinion](https://github.com/rust-lang/rust/issues/105999#issuecomment-1363429848) that it's a perfectly valid way to parse argv:

> To the best of my knowledge, it's completely valid to parse argv by stopping at the first NULL; if an argument parser is replacing an argument with NULL, it should not expect that arguments after that point will be seen. I've seen plenty of C programs that just do `for (char **arg = argv; *arg; arg++)` or equivalent.

Argument parsers mutating argv sounds like a very dicey situation and Rust's std should play it safe.

@bors r+

**Comment by sdroege** [comment]
Can this also be backported to beta so it ends up in 1.68?

**Comment by RalfJung** [comment]
We usually only backport regression fixes I think. @Mark-Simulacrum do you think this qualifies for a backport?

**Comment by Mark-Simulacrum** [comment]
It's up to T-libs, but I would lean towards no. We land bugfixes routinely and don't typically backport them. This is a soundness fix I guess, but only in edge cases, so I don't think it should merit special consideration.

**Comment by RalfJung** [comment]
@rust-lang/libs should this be backported to beta?

**Comment by ChrisDenton** [comment]
I've added the "beta-nominated" label to make sure it gets discussed. This will indeed need t-libs sign off before being accepted (or not).

**Pull request merged**